src/node/transaction.cpp

```diff
@@ -39,9 +39,19 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
         // Transaction is not already in the mempool.
         TxValidationState state;
         CAmount fee{0};
-        if (AcceptToMemoryPool(*node.mempool, state, tx,
-                nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee)) {
-            if (max_tx_fee && fee > max_tx_fee) {
+        if (max_tx_fee) {
+            if (!GetTransactionFee(*node.mempool, state, tx, &fee)) {
+                err_string = state.ToString();
+                if (state.IsInvalid()) {
+                    if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
+                        return TransactionError::MISSING_INPUTS;
+                    }
+                    return TransactionError::MEMPOOL_REJECTED;
+                } else {
+                    return TransactionError::MEMPOOL_ERROR;
+                }
```

Comment:
It's a shame to duplicate the code a few lines below. If we used `GetTransactionFee()` _just_ to get the transaction fee (and not return any validation state) and relied on ATMP to return validation state, we wouldn't have this code duplication.

Reply:
I made a helper function for this so it doesn't duplicate the code
The question of "if GetFee fails, still run ATMP or fail immediately?" is still up in the air for me...

Reply:
@gzhao408 What is the benefit of running ATMP if `GetTransactionFee()` fails?  It seems to me that if you get a failure, you should abort -- particularly because the way it's being used here is to prevent an absurd fee from getting through.  If a transient failure prevented the fee from being calculated (say, a race condition with an input being confirmed in a block or arriving in the mempool from another transaction), then continuing after the fee check fails to return a value risks actually sending out a transaction in the call to ATMP.

@jnewbery I think `GetTransactionFee()` should return a reason for failure.  For instance, I could imagine that we might want to expose a `getrawtransactionfee` RPC where we just return the fee of a rawtransaction, and being able to return a meaningful error for why the fee couldn't be calculated seems useful in that context.

Reply:
@sdaftuar I definitely agree that if `GetTransactionFee` fails, we shouldn't try to submit the transaction. My issue with using `GetTransactionFee` to return a `TxValidationResult` is that, in my opinion, ATMP should be the authority on validation results.

At the end of the day, the fee is not meaningful unless validation succeeds. If `GetTransactionFee` fails due to the transaction being invalid, yes it fails, but it's more useful to return the validation error than "sorry, couldn't get the fee." But I don't think `GetTransactionFee` should be relied upon to return a validation error. It would make sense to me to run a `test_accept` of ATMP to get the proper validation error if `GetTransactionFee` fails. But then... why not just dry-run ATMP?

Reply:
Also @sdaftuar the `testmempoolaccept` RPC already returns the fee for a raw transaction (#19940)!